### ISSUE TITLE
Decide on using Prefer or Prefer-Push header

### ISSUE THREAD

**Issue by evert** [evert/push-please#3]
Benefits of using a new header is that it might make the syntax (for some of the formats) simpler due to less escaping. Another benefit is that it slightly reduces friction as implementors don't need a more complex parsers/understand the `Prefer` header structure.

Benefits of using the existing Prefer header is that the header conveys 'optional preference' in a general way.

Not sure if I did the argument of `Prefer` justice. Feel free to edit/amend.

**Comment by pezra** [comment]
I prefer `Prefer`. 😄

**Comment by gabesullice** [comment]
@pezra, do you feel that there are other benefits to `Prefer` that @evert did not cover?

If we do not reuse the `Prefer` header, I think we should consider a name without "Prefer" in it because it raises the obvious question: "why was `Prefer` insufficient?"

**Comment by evert** [comment]
@hibaymj says:

> Preventing bifurcation. It’s the standard way to express preferences the server can ignore.
The benefits of further pushing Prefer over multiple bespoke headers to accomplish the same thing are the exact same benefits to 
coalescing on standard API formats and mediaTypes to prevent bespoke implementations. Since the standard exists and we’re talking about pushing another header as a standard, it would seem the onus is on those suggesting a new header to argue why the standard header wouldn’t work.

**Comment by evert** [comment]
From the chat, written by Mike Hibay

**Comment by evert** [comment]
> Since the standard exists and we’re talking about pushing another header as a standard, it would seem the onus is on those suggesting a new header to argue why the standard header wouldn’t work.

My 2ct:  I rather ask folks to parse 1 format instead of 2. Especially when using single-depth pushes, i want to just point people to the structured-header parser. I'm also not too worried about the drawbacks of having a new header. I don't believe that the HTTP standards are going in a direction where every 'optional' feature all has to be tested under a potentially complex `Prefer` header. 

Anyway, either version of this is _definitely_ not a blocker to me, which is why I asked HTTP-WG for input.  I want to give this the highest chance of success, so if that group leans to either `Prefer` or `Prefer-Push` I will 100% jump on that bandwagon.

**Comment by petejohanson** [comment]
Someone had noted that the `Preference-Applied` response header could be leveraged if `Prefer` is used https://tools.ietf.org/html/rfc7240#section-3

**Comment by pezra** [comment]
@gabesullice said,
> @pezra, do you feel that there are other benefits to `Prefer` 

My thinking is similar to @hibaymj's. I prefer fewer standards. `Prefer` already exists and is perfectly serviceable for this purpose. 

@evert said,
> i want to just point people to the structured-header parser

We can use `Prefer` *and* a standard structured-header parser. Consider `Prefer: push` and `Prefer: push;rel=item`.

This is not a blocker for me either, i just see no reason to introduce another header.

**Comment by evert** [comment]
Hi @pezra,

How would you turn this header:

    Prefer-Push: rel, "http://example.org/rels/foo"

into a structured-headers compatible value?

**Comment by evert** [comment]
Actually, `Prefer: push;rel=item.` is not even valid for `Prefer`. See

https://tools.ietf.org/html/rfc7240#section-2

It's quite limited in what it permits.

**Comment by handrews** [comment]
@evert how is that not valid?  Both the preference and its parameters may be either single strings or name=value pairs (where single strings are considered to be names with an empty string value)

**Comment by evert** [comment]
You're right, I misread. Forget that last comment. The one before it still stands.
